File: leetcode/Evaluate_Reverse_Polish_Notation.cpp
class Solution {
public:
    int evalRPN(vector<string> &tokens) {
        if(tokens.size() == 0) throw runtime_error("empty input!");
        
        stack<int> st;
        for(int i = 0; i < tokens.size(); i++){
            if(tokens[i] == "+" || tokens[i] == "-" || tokens[i] == "/" || tokens[i] == "*"){
                opExec(st, tokens[i]);        
            }else{
                st.push(stoi(tokens[i]));
            }
        }

        if(st.empty()) throw runtime_error("wrong expr!");
        else return st.top();
    }
private:
    void opExec(stack<int> &st, string &op){
        int a, b;
        if(st.empty()) throw runtime_error("wrong expr");
        a = st.top();
        st.pop();
                
        if(st.empty()) throw runtime_error("wrong expr");
        b = st.top();
        st.pop();
    
        if(op == "+") st.push(b+a);
        if(op == "-") st.push(b-a);
        if(op == "/") st.push(b/a);
        if(op == "*") st.push(b*a);
    }
};

File: leetcode/java/EvaluateReversePolishNotation.java
class Solution {
    public int evalRPN(String[] tokens) {
        if(tokens == null || tokens.length == 0) throw new IllegalArgumentException("x");
        Deque<Integer> st = new ArrayDeque<>();
        
        for(String t : tokens) {
            if(isOp(t)) {
                int x = st.pop();
                int y = st.pop();
                st.push(compute(x, y, t));
            }else {
                st.push(Integer.valueOf(t));   
            }
        }
        return st.pop();
    }
    
    private boolean isOp(String st) {
        if(st.equals("+") ||st.equals("-") || st.equals("/") || st.equals("*") ) return true;
        return false;
    }
    
    private int compute(int x, int y, String op) {
        if(op.equals("+")) return x+y;
        else if(op.equals("-")) return y-x;
        else if(op.equals("*")) return x*y;
        else return y/x;
    }
}
